File: typescript/src/services/EncryptionService.ts
import { createCipheriv, createDecipheriv, randomBytes } from 'crypto'
import { IEncryptionService } from '../interfaces/IEncryptionService'

/**
 * Serviço de criptografia implementando IEncryptionService.
 * Gerencia a criptografia e descriptografia de dados sensíveis.
 */
export class EncryptionService implements IEncryptionService {
  private readonly key: Buffer
  private readonly algorithm = 'aes-256-gcm'

  constructor() {
    const encryptionKey = process.env.ENCRYPTION_KEY
    if (!encryptionKey) {
      throw new Error('ENCRYPTION_KEY environment variable is required')
    }
    this.key = Buffer.from(encryptionKey, 'base64')
  }

  /**
   * Criptografa uma string.
   * @param data String a ser criptografada
   * @returns Buffer com dados criptografados
   */
  encrypt(data: string): Buffer {
    try {
      if (!data) {
        return Buffer.from('')
      }

      const iv = randomBytes(12)
      const cipher = createCipheriv(this.algorithm, this.key, iv)

      const encrypted = Buffer.concat([
        cipher.update(data, 'utf8'),
        cipher.final()
      ])

      const authTag = cipher.getAuthTag()

      // Combina IV + AuthTag + Dados Criptografados
      return Buffer.concat([iv, authTag, encrypted])
    } catch (error) {
      console.error('Erro ao criptografar dados:', error)
      throw error
    }
  }

  /**
   * Descriptografa uma string.
   * @param encryptedData Buffer com dados criptografados
   * @returns String original descriptografada
   */
  decrypt(encryptedData: Buffer): string {
    try {
      if (!encryptedData.length) {
        return ''
      }

      // Extrai IV, AuthTag e dados criptografados do buffer
      const iv = encryptedData.subarray(0, 12)
      const authTag = encryptedData.subarray(12, 28)
      const data = encryptedData.subarray(28)

      const decipher = createDecipheriv(this.algorithm, this.key, iv)
      decipher.setAuthTag(authTag)

      const decrypted = Buffer.concat([decipher.update(data), decipher.final()])

      return decrypted.toString('utf8')
    } catch (error) {
      console.error('Erro ao descriptografar dados:', error)
      throw error
    }
  }
}

File: python/src/services/encryption_service.py
"""
Serviço de criptografia implementando IEncryptionService.

Gerencia a criptografia e descriptografia de dados sensíveis.
"""

import os

from cryptography.fernet import Fernet

from ..interfaces.encryption_service_interface import IEncryptionService
from ..utils.log import log


class EncryptionService(IEncryptionService):
    """Serviço de criptografia usando Fernet."""

    def __init__(self):
        """
        Inicializa o serviço de criptografia.

        Args:
            encryption_key: Chave de criptografia. Se None, busca em CHAVE_CRIPTOGRAFIA
        """
        self._key = str(os.environ.get("CHAVE_CRIPTOGRAFIA")).encode()
        self._fernet = Fernet(self._key)

    def encrypt(self, data: str) -> bytes:
        """
        Criptografa uma string.

        Args:
            data: String a ser criptografada

        Returns:
            String criptografada em base64
        """
        try:
            if not data:
                return b""

            return self._fernet.encrypt(data.encode())

        except Exception as e:
            log.error(f"Erro ao criptografar dados: {e}")
            raise

    def decrypt(self, encrypted_data: str) -> str:
        """
        Descriptografa uma string.

        Args:
            encrypted_data: String criptografada em base64

        Returns:
            String original descriptografada
        """
        try:
            if not encrypted_data:
                return encrypted_data

            return self._fernet.decrypt(encrypted_data).decode()

        except Exception as e:
            log.error(f"Erro ao descriptografar dados: {e}")
            raise
